File: TestLeetCode/Graph/GraphIsTree.cs
public class ValidTreeSolution {
    public bool ValidTree(int n, int[][] edges) {
        if (n == 0) {
            return true;
        }

        // Step 1: Create adjacency list
        var adj = new Dictionary<int, List<int>>();
        for (int i = 0; i < n; i++) {
            adj[i] = new List<int>();
        }
        
        foreach (var edge in edges) {
            int n1 = edge[0], n2 = edge[1];
            adj[n1].Add(n2);
            adj[n2].Add(n1);
        }

        var visit = new HashSet<int>();

        // Step 2: Depth-First Search (DFS) to check for cycles and connectivity
        bool Dfs(int node, int parent) {
            if (visit.Contains(node)) {
                return false;
            }

            visit.Add(node);
            foreach (var neighbor in adj[node]) {
                if (neighbor == parent) {
                    continue;
                }
                if (!Dfs(neighbor, node)) {
                    return false;
                }
            }
            return true;
        }

        // Check if the graph is connected and acyclic
        return Dfs(0, -1) && visit.Count == n;
    }
}

File: python/graph/graph_is_tree.py
class ValidTree:
    def valid_tree(self, n: int, edges: list[list[int]]) -> bool:
        if n == 0: return True
        if len(edges) != n - 1: return False
        graph = [[] for _ in range(n)]
        for left, right in edges: graph[left].append(right); graph[right].append(left)
        seen = set()
        def visit(node: int, parent: int) -> None:
            seen.add(node)
            for neighbor in graph[node]:
                if neighbor != parent and neighbor not in seen: visit(neighbor, node)
        visit(0, -1)
        return len(seen) == n
